tick/optim/solver/src/saga.cpp

```diff
@@ -0,0 +1,182 @@
+// License: BSD 3 clause
+
+#include "model_labels_features.h"
+#include "saga.h"
+
+SAGA::SAGA(ulong epoch_size,
+           double tol,
+           RandType rand_type,
+           double step,
+           int seed,
+           VarianceReductionMethod variance_reduction)
+    : StoSolver(epoch_size, tol, rand_type, seed),
+      step(step), variance_reduction(variance_reduction),
+      solver_ready(false),
+      ready_step_corrections(false) {}
+
+void SAGA::set_model(ModelPtr model) {
+  // TODO: model should be a child of ModelGeneralizedLinear
+  StoSolver::set_model(model);
+  ready_step_corrections = false;
+  solver_ready = false;
+}
+
+void SAGA::initialize_solver() {
+  ulong n_samples = model->get_n_samples();
+  gradients_memory = ArrayDouble(n_samples);
+  gradients_memory.fill(0);
+  gradients_average = ArrayDouble(model->get_n_coeffs());
+  gradients_average.fill(0);
+  solver_ready = true;
+}
+
+void SAGA::prepare_solve() {
+  // The point where we compute the full gradient for variance reduction is the
+  // new iterate obtained at the previous epoch
+  if (!solver_ready) {
+    initialize_solver();
+  }
+  next_iterate = iterate;
+  if ((model->is_sparse()) && (prox->is_separable())) {
+    if (!ready_step_corrections) {
+      compute_step_corrections();
+    }
+  }
+  rand_index = 0;
+  if (variance_reduction == VarianceReductionMethod::Random ||
+      variance_reduction == VarianceReductionMethod::Average) {
+    next_iterate.init_to_zero();
+  }
+  if (variance_reduction == VarianceReductionMethod::Random) {
+    rand_index = rand_unif(epoch_size);
+  }
+}
+
+void SAGA::solve() {
+  prepare_solve();
+  bool use_intercept = model->use_intercept();
+  ulong n_features = model->get_n_features();
+  if ((model->is_sparse()) && (prox->is_separable())) {
+    solve_sparse_proba_updates(use_intercept, n_features);
+  } else {
+    solve_dense(use_intercept, n_features);
+  }
```

Comment:
not a very big deal but why don't you call 
```
bool use_intercept = model->use_intercept();
ulong n_features = model->get_n_features();
```
inside `solve_sparse_proba_updates` and `solve_dense` rather than passing them as parameters. It's strange to me to pass only half as parameters